python/lsst/ts/cbp/component.py

```diff
@@ -136,29 +143,63 @@ def generate_mask_info(self):
         mask_dict["9"].name = "Unknown"
         self.masks = types.SimpleNamespace(**mask_dict)
 
-    def get_in_position_status(self):
-        """Get the in position status of each actuator."""
-        self.encoder_in_position.azimuth = (
+    def get_in_position_status(self, changed=True, move_commanded=[]):
+        """Get the in position status of each actuator.
+
+        Also calls a method to publish the inPosition event if the status of
+        any actuator has changed since the previous call.
+        """
+        changed = changed
```

Comment:
I suggest that you store in_position, target position and current position **only** in the SAL topics, instead of keeping a duplicate of this information in your own custom classes.

This eliminates a lot of custom code and also eliminates any risk of the two copies of the information getting out of synch with each other.

Here is what that looks like:

1) The AxisInPosition class and its cousins disappear.

2) Use accessors to get the data from the topics:
```
@property
def in_position(self):
    return self.csc.evt_inPosition.data

@property
def target(self):
    return self.csc.evt_target.data

@propery
def azimuth(self):
    return self.csc.tel_CBP_azimuth.azimuth

@property
def elevation(self):
    return self.csc.tel_CBP_elevation.elevation

... etc. for the other axes.
```

3) In the move commands, explicitly set the fields false for the axes being moved, once you know that the move has successfully started. For example:
```
def move_azimuth(self, azimuth):
    # ... do all your usual stuff here, and if the move command succeeds
    self.csc.evt_inPosition.set_put(azimuth=False)
```

4) This entire method (and its cousins that set your other custom classes) are much simpler; just set the data once and you're done. Here is the rewritten version of this method, which I also renamed because it doesn't get anything; it just sets. It is dramatically simpler and yet it does everything needed:
```
def update_in_position(self):
    """Update the in position status of each actuator,
    based on the most recently read encoder data.

    Returns
    --------
    did_change : `bool`
        True if anything changed (and so the event was published)
    """
    return self.csc.evt_inPosition.set_put(
        azimuth=abs(self.azimuth - self.target.azimuth) < self.error_tolerance,
        elevation=abs(self.elevation - self.target.altitude) < self.error_tolerance,
        mask=self.mask == self.target.mask,
        mask_rotation=(
            abs(self.mask_rotation - self.target.mask_rotation) < self.error_tolerance
        ),
        focus=abs(self.focus - self.target.focus) < self.error_tolerance,
    )
```
If your code ever needs to force this even to be published then you can add a `force_output` argument that defaults to False. `force_output` is a valid keyword argument for the `set_put` method of event topics (but not telemetry topics because telemetry set_put *always* outputs the topic).

Reply:
I am concerned that your error tolerance is the same for all axes. Is that safe? I would expect it to work for azimuth and elevation (since the encoder resolution and motion accuracy is probably similar for both) but I am concerned about mask rotation and focus.

If you have one tolerance for each axis (with a value of 0 for mask) you could even automate the in_position code, which nicely avoid typos but is a bit harder to read (you need `<=` instead of `<` if a tolerance can be 0):
```
in_position_kwargs = {
    axis: abs(
        getattr(self, axis) - getattr(self.target, axis)
    )
    <= getattr(self.tolerance, axis)
    for axis in self.axis_names
}
self.csc.evt_inPosition.set_put(**in_position_kwargs)
```